pkg/api/certificate_cache.go

```diff
@@ -0,0 +1,205 @@
+	ID         string     `json:"id"`
+}
+
+// Attributes contains the specific metadata for a TLS certificate,
+// such as its Common Name (CN), user-defined name, issuer,
+// expiration date (NotAfter), and serial number (SN).
+type Attributes struct {
+	CN       string `json:"issued_to"`
+	Name     string `json:"name"`
+	Issuer   string `json:"issuer"`
+	NotAfter string `json:"not_after"`
+	SN       string `json:"serial_number"`
+}
+
+// CertificateCache polls api.fastly.com/tls/certificates and maintains a local cache
+// of the returned metadata. That information is exposed as Prometheus metrics.
+type CertificateCache struct {
+	client  HTTPClient
+	token   string
+	enabled bool
+	logger  log.Logger
+
+	mtx   sync.Mutex
+	certs []Certificate
+}
+
+// NewCertificateCache returns an empty cache of certificates metadata. Use the
+// Refresh method to update the cache.
+func NewCertificateCache(client HTTPClient, token string, enabled bool, logger log.Logger) *CertificateCache {
+	return &CertificateCache{
+		client:  client,
+		token:   token,
+		enabled: enabled,
+		logger:  logger,
+	}
+}
+
+// Refresh the cache with metadata retreived from the Fastly API.
+func (c *CertificateCache) Refresh(ctx context.Context) error {
+	if !c.enabled {
+		return nil
+	}
+	begin := time.Now()
+
+	var (
+		uri       = fmt.Sprintf("https://api.fastly.com/tls/certificates?page%%5Bnumber%%5D=1&page%%5Bsize%%5D=%d&sort=created_at", maxCertificatesPageSize)
+		nextCerts = []Certificate{}
+		total     = 0
+	)
+	for {
+		req, err := http.NewRequestWithContext(ctx, "GET", uri, nil)
+		if err != nil {
+			return fmt.Errorf("error constructing API certificates request: %w", err)
+		}
+
+		req.Header.Set("Fastly-Key", c.token)
+		req.Header.Set("Accept", "application/json")
+		resp, err := c.client.Do(req)
+		if err != nil {
+			return fmt.Errorf("error executing API certificates request: %w", err)
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode != http.StatusOK {
+			// disable if we get a 403 and c.certs is nil (only true on first run)
+			if resp.StatusCode == http.StatusForbidden && c.certs == nil {
+				c.enabled = false
+			}
+			return NewError(resp)
+		}
+
+		var response CertificateResponse
+		if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
+			return fmt.Errorf("error decoding API certificates response: %w", err)
+		}
+
+		nextCerts = append(nextCerts, response.Certificates...)
+		total += len(response.Certificates)
+
+		next, err := getNextCertificateLink(response)
+		if err != nil {
+			break
+		}
+
+		uri = next.String()
+	}
+
+	level.Debug(c.logger).Log(
+		"certificate_refresh_took", time.Since(begin),
+		"total_certificate_count", total,
+	)
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	c.certs = nextCerts
+
+	return nil
+}
+
+// Certificates returns a copy of the currently cached certificates.
+func (c *CertificateCache) Certificates() []Certificate {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	certs := c.certs
```

Comment:
Probably be better as 
```suggestion
	certs := c.certs[:]
```